admin/server/github.go

```diff
@@ -237,6 +237,95 @@ func (s *Server) GetGithubRepoStatus(ctx context.Context, req *adminv1.GetGithub
 	return res, nil
 }
 
+func (s *Server) ListGithubUserRepos(ctx context.Context, req *adminv1.ListGithubUserReposRequest) (*adminv1.ListGithubUserReposResponse, error) {
+	claims := auth.GetClaims(ctx)
+	if claims.OwnerType() != auth.OwnerTypeUser {
+		return nil, status.Error(codes.Unauthenticated, "not authenticated")
+	}
+
+	userID := claims.OwnerID()
+	user, err := s.admin.DB.FindUser(ctx, userID)
+	if err != nil {
+		return nil, err
+	}
+
+	// user has not authorized github app
+	if user.GithubUsername == "" {
+		return nil, status.Error(codes.Unauthenticated, "not authenticated")
+	}
+
+	token, refreshToken, err := s.userAccessToken(ctx, user.GithubRefreshToken)
+	if err != nil {
+		return nil, status.Error(codes.Internal, err.Error())
+	}
+
+	// refresh token changes after using it for getting a new token
+	// so saving the updated refresh token
+	_, err = s.admin.DB.UpdateUser(ctx, claims.OwnerID(), &database.UpdateUserOptions{
+		DisplayName:         user.DisplayName,
+		PhotoURL:            user.PhotoURL,
+		GithubUsername:      user.GithubUsername,
+		GithubRefreshToken:  refreshToken,
+		QuotaSingleuserOrgs: user.QuotaSingleuserOrgs,
+		PreferenceTimeZone:  user.PreferenceTimeZone,
+	})
+	if err != nil {
+		return nil, fmt.Errorf("failed to update user: %w", err)
+	}
+
+	client := github.NewTokenClient(ctx, token)
+
+	// use a client with user's token to get installations
+	installations, _, err := client.Apps.ListUserInstallations(ctx, &github.ListOptions{
+		Page:    0,
+		PerPage: 0,
+	})
+	if err != nil {
+		return nil, status.Error(codes.Internal, err.Error())
+	}
+
+	repos := make([]*adminv1.ListGithubUserReposResponse_Repo, 0)
+	for _, installation := range installations {
+		instToken, _, err := s.admin.Github.AppClient().Apps.CreateInstallationToken(ctx, *installation.ID, nil)
+		if err != nil {
+			return nil, status.Error(codes.Internal, err.Error())
+		}
+
+		// use the installation token to fetch the repos accessible to it
+		instClient := github.NewTokenClient(ctx, *instToken.Token)
+
+		req, err := instClient.NewRequest("GET", *installation.RepositoriesURL, nil)
```

Comment:
Can directly use `instClient.Apps.ListUserRepos(ctx, *installation.ID, &github.ListOptions{})`

Reply:
I tried that 1st. Got `403 Resource not accessible by integration`

Reply:
Okay just realised that it works with [user access token](https://docs.github.com/en/rest/apps/installations?apiVersion=2022-11-28#list-repositories-accessible-to-the-user-access-token--fine-grained-access-tokens). In that case the above `client` itself will work and we don't need to generate a new client.

Reply:
Also shouldn't the other request `instClient.NewRequest("GET", *installation.RepositoriesURL, nil)`  also respond with `403` unless the underlying API does something else.

Reply:
Ah i was missing something earlier i think. I get the response now with `client`